presto-main/src/main/java/com/facebook/presto/operator/aggregation/fixedhistogram/FixedDoubleBreakdownHistogram.java

```diff
@@ -0,0 +1,266 @@
+        this.weights = new double[0];
+        this.counts = new long[0];
+    }
+
+    public FixedDoubleBreakdownHistogram(SliceInput input)
+    {
+        this.bucketCount = input.readInt();
+        this.min = input.readDouble();
+        this.max = input.readDouble();
+        verifyParams(bucketCount, min, max);
+        int size = input.readInt();
+        this.indices = new int[size];
+        this.weights = new double[size];
+        this.counts = new long[size];
+        input.readBytes(
+                Slices.wrappedIntArray(indices),
+                size * SizeOf.SIZE_OF_INT);
+        input.readBytes(
+                Slices.wrappedDoubleArray(weights),
+                size * SizeOf.SIZE_OF_DOUBLE);
+        input.readBytes(
+                Slices.wrappedLongArray(counts),
+                size * SizeOf.SIZE_OF_LONG);
+    }
+
+    protected FixedDoubleBreakdownHistogram(FixedDoubleBreakdownHistogram other)
+    {
+        this.bucketCount = other.bucketCount;
+        this.min = other.min;
+        this.max = other.max;
+        verifyParams(bucketCount, min, max);
+        this.indices = new int[other.indices.length];
+        this.weights = new double[other.weights.length];
+        this.counts = new long[other.counts.length];
+        for (int i = 0; i < other.indices.length; ++i) {
+            this.indices[i] = other.indices[i];
+            this.weights[i] = other.weights[i];
+            this.counts[i] = other.counts[i];
+        }
+    }
+
+    public int getBucketCount()
+    {
+        return bucketCount;
+    }
+
+    public double getMin()
+    {
+        return min;
+    }
+
+    public double getMax()
+    {
+        return max;
+    }
+
+    public double getWidth()
+    {
+        return (max - min) / bucketCount;
+    }
+
+    public int getRequiredBytesForSerialization()
+    {
+        return
+                SizeOf.SIZE_OF_INT + // bucketCount
+                        2 * SizeOf.SIZE_OF_DOUBLE + // min, max
+                        SizeOf.SIZE_OF_INT + // size
+                        indices.length * (SizeOf.SIZE_OF_INT + SizeOf.SIZE_OF_DOUBLE + SizeOf.SIZE_OF_LONG);
+    }
+
+    public void serialize(SliceOutput out)
+    {
+        out.appendInt(bucketCount);
+        out.appendDouble(min);
+        out.appendDouble(max);
+        out.appendInt(indices.length);
+        int i = 0;
+        for (i = 0; i < bucketCount; ++i) {
+            out.appendDouble(indices[i]);
+        }
+        for (i = 0; i < bucketCount; ++i) {
+            out.appendDouble(weights[i]);
+        }
+        for (i = 0; i < bucketCount; ++i) {
+            out.appendDouble(counts[i]);
+        }
+    }
+
+    public void add(double value)
+    {
+        add(value, 1.0);
+    }
+
+    public void add(double value, double weight)
+    {
+        add(value, weight, 1);
+    }
+
+    public void add(double value, double weight, long count)
+    {
+        int bucketIndex = getIndexForValue(bucketCount, min, max, value);
```

Comment:
Changed the name to `foundIndex`. Is that OK?

Reply:
Sounds good!

Reply:
Hey, In retrospect, I'm not sure about your comment (thought you meant something else before). Why isn't this the actual bucket index?

Reply:
So here's how I understand it:

We call `getIndexForValue`.  This calculates the bucket based on its position presuming that each bucket has only one bucket-weight combo.  But if there are multiple weights, this will not be correct, because it doesn't take into account the fact that an additional weight will shift all the buckets on the right by one.  So in reality it seems to be a naive guess, which probably works most of the time since usually we'll have constant weights (according to my understanding).  Did I understand correctly?

Reply:
Ah, got it now. I think the naming confused you a bit (and have a suggestion). 

There are two indices: the logical bucket index, and the index where this bucket/weight combination resides. Calling `getIndexForValue` returns the _exact_ logical bucket index, but, as you pointed out, it doesn't say much about the location where this entry resides. 

To clarify things, I'm renaming `bucketIndex` to `logicalBucketIndex`. WDYT?

Reply:
This is much better, thanks!